### ISSUE TITLE
[naming-convention] Allow multiple underscore

### ISSUE THREAD

**Issue by jhnns** [typescript-eslint/typescript-eslint#1712]
**Repro**

Hey ho 👋, this is a feature request. What do you think of allowing multiple leading or trailing underscores? There are some libraries that use double underscore (`__`) to avoid name clashes. Probably the most popular one is GraphQL where you can specify a `__typename` (see [here](https://graphql.org/learn/queries/)).

Personally, I want to use camelCase or PascalCase and I'm ok with people pre- or postfixing names with underscores. We could introduce a 4th value for the `leadingUnderscore` and `trailingUnderscore` option that is called `allowMultiple`, like this:

```JSON
{
  "rules": {
    "@typescript-eslint/naming-convention": ["warn", {
        "format": ["camelCase", "PascalCase", "UPPER_CASE"],
        "leadingUnderscore": "allowMultiple",
        "selector": "default",
        "trailingUnderscore": "allowMultiple"
    }]
  }
}
```

**Comment by bradzacher** [comment]
It's a pattern people use for protected members as well.
Happy if you'd like to add the option!

**Comment by felixfbecker** [comment]
Also React's `dangerouslySetInnerHTML={{ __html: "" }}`

**Comment by ScottAwesome** [comment]
I want to comment here as well, as I would love to see this. I tried doing a custom regex filter, so if anything started with double leading underscores, it wouldn't error:

```js
const propertySetting = selector: 'property',
                format: ['camelCase', 'PascalCase', 'UPPER_CASE'],
                filter: {
                    regex: '^(__|_)([0-9A-Za-z]+)$',
                    match: false,
                };
```


No such luck, giving me errors. 

If anyone has a secret sauce here that does work, please share!

Also, another option to think about is allowing optionalPrefixes instead? if I do prefix with a blank string, eslint throws.

**Comment by joh-klein** [comment]
@ScottAwesome I got your filter to work, when the ONLY rule was the one you posted (`propertySetting`). When there are other rules that fit (in my case I had another one with `selector: 'default'`), it will apply those rules again – and possibly fail.

**Comment by BorisZubchenko** [comment]
For those looking for a `__typename` solution:
```
{
    "selector": "property",
    "filter": "__typename",
    "format": null
 }
```
Good:
```
console.log({
  __typename: 'Type'
});
```

**Comment by ScottAwesome** [comment]
> For those looking for a `__typename` solution:
> 
> ```
> {
>     "selector": "property",
>     "filter": "__typename",
>     "format": null
>  }
> ```
> 
> Good:
> 
> ```
> console.log({
>   __typename: 'Type'
> });
> ```

This doesn’t solve for the fact I only want certain formats otherwise though, does it? See my previous example where it won’t work if other rules are being applied

**Comment by ab-pm** [comment]
What I found to be working is a `prefix`:
```js
{
	selector: 'property',
	prefix: ['__'],
	format: ['camelCase']
},
```

However, for some reason (not sure if it's a bug) I had to restrict the normal `property` rule to not match the underscored properties, as it would complain otherwise:
```js
{
	selector: 'property',
	filter: '^(?!__)',
	format: […]
},
```
